dns-server/src/dns_server.rs

```diff
@@ -22,191 +31,244 @@ use trust_dns_proto::rr::{Name, Record};
+                    )
+                    .await
+                }
+                RequestError::ServFail(_) => {
+                    respond_servfail(&request, rb_servfail, &header).await
+                }
+            };
+        }
     }
+}
 
-    let name = mr.query().original().name().clone();
-    let key = name.to_string();
-    let key = key.trim_end_matches('.');
-
-    let rb = MessageResponseBuilder::from_message_request(&mr);
-
-    let bits = match db.get(key.as_bytes()) {
-        Ok(Some(bits)) => bits,
-
-        // If no record is found bail with NXDOMAIN.
-        Ok(None) => {
-            respond_nxdomain(&log, socket, src, rb, header, &mr).await;
-            return;
-        }
+/// Describes how to respond to a particular request failure
+#[derive(Debug, Error)]
+enum RequestError {
+    #[error("NXDOMAIN: {0:#}")]
+    NxDomain(#[source] QueryError),
+    #[error("SERVFAIL: {0:#}")]
+    ServFail(#[source] anyhow::Error),
+}
 
-        // If we encountered an error bail with SERVFAIL.
-        Err(e) => {
-            error!(log, "db get: {}", e);
-            nack(&log, &mr, &socket, &header, &src).await;
-            return;
+impl From<QueryError> for RequestError {
+    fn from(source: QueryError) -> Self {
+        match &source {
+            QueryError::NoName(_) => RequestError::NxDomain(source),
+            // Bail with servfail when this query is for a zone that we don't
+            // own (and other server-side failures) so that resolvers will look
+            // to other DNS servers for this query.
+            QueryError::NoZone(_)
+            | QueryError::QueryFail(_)
+            | QueryError::ParseFail(_) => RequestError::ServFail(source.into()),
         }
-    };
-
-    let records: Vec<crate::dns_data::DnsRecord> =
-        match serde_json::from_slice(bits.as_ref()) {
-            Ok(r) => r,
-            Err(e) => {
-                error!(log, "deserialize record: {}", e);
-                nack(&log, &mr, &socket, &header, &src).await;
-                return;
-            }
-        };
-
-    if records.is_empty() {
-        error!(log, "No records found for {}", key);
-        respond_nxdomain(&log, socket, src, rb, header, &mr).await;
-        return;
     }
+}
+
+/// Handle a well-formed, decoded DNS query
+async fn handle_dns_message(
+    request: &Request,
+    mr: &MessageRequest,
+) -> Result<(), RequestError> {
+    let log = &request.log;
+    let store = &request.store;
+    debug!(&log, "message_request"; "mr" => #?mr);
 
-    let mut response_records: Vec<Record> = vec![];
-    for record in &records {
-        let resp = match record {
+    let header = Header::response_from_request(mr.header());
+    let query = mr.query();
+    let name = query.original().name().clone();
+    let records = store.query(mr)?;
+    let rb = MessageResponseBuilder::from_message_request(mr);
+    let response_records = records
+        .into_iter()
+        .map(|record| match record {
```

Comment:
For external DNS, we'll need to handle `A` and `NS` records too.

Reply:
💯  I'd like to put this in a follow-up PR since this one is already so large.

Reply:
Sounds good.